src/routes/session.ts

```diff
@@ -5,26 +5,69 @@ import {
 	comparePassword,
 	runInvalidPasswordComparison,
 } from '../platform/auth';
-import { sbvrUtils } from '../platform';
+import { resinApi, root, sbvrUtils } from '../platform';
 import { captureException, handleHttpErrors } from '../platform/errors';
 import { resetCounter } from '../lib/rate-limiting';
 import { RequestHandler } from 'express';
 import { SetupOptions } from '..';
 
 const { BadRequestError, NotFoundError } = sbvrUtils;
 
-export const whoami: RequestHandler = (req, res) =>
-	getUser(req)
-		.then(user =>
-			res.send({ id: user.id, username: user.username, email: user.email }),
-		)
-		.catch(err => {
-			if (handleHttpErrors(req, res, err)) {
-				return;
+export const whoami: RequestHandler = async (req, res) => {
+	try {
+		let user: AnyObject = await getUser(req);
+
+		// If the user email is available, then the record was successfully
+		// populated by the passport middleware
+		if (user.email == null) {
+			// If we get to this point, then the request is most likely being made
+			// using a scoped access token.
+			// First we check to see if this token can access the user resource,
+			// then if it can, we retrieve user document that corresponds to the
+			// `actor` value on the token using root privileges.
+			const results = await resinApi.get({
+				resource: 'user',
+				passthrough: { req },
+				options: {
+					$filter: {
+						actor: user.actor,
+					},
+					$select: 'id',
+				},
+			});
+
+			// If the count is 0, then this token doesn't have access to the
+			// user resource and we should just continue with whatever was
+			// provided from the `getUser` call above.
+			if (results.length > 0) {
+				// If we reach this step, then the token has access and we can
+				// retrieve the user document in full
+				const id = results[0].id;
+
+				user = (await resinApi.get({
+					resource: 'user',
+					passthrough: { req: root },
+					id,
+					options: {
+						$select: ['id', 'username', 'email'],
+					},
+				})) as AnyObject;
```

Comment:
```suggestion
				})) as Promise<Pick<DbUser, 'id' | 'username' | 'email'>>;
```
Where the DbUser is `import { User as DbUser } from '../platform/models';`

Reply:
Nice!